origin-js/src/models/offer.js

```diff
@@ -3,23 +3,52 @@
 //
 export class Offer {
   /**
-   * An Offer is constructed based on its on-chain and off-chain data.
-   * @param {string} offerId - Unique offer ID.
-   * @param {string} listingId - Unique listing ID.
-   * @param {Object} chainOffer - Offer data from the blockchain.
-   * @param {Object} ipfsOffer - Offer data from IPFS.
+   * The same offer object constructed by origin-js and discovery server. Ideally we would have
+   * separate constructors but JavaScript does not support that. For that reason this 2 mode 
+   * constructor paradigm is used
+   *
+   * @param {Object} originJsData - Constructor argument used by marketplace resource
+   * - {string} offerId - Unique offer ID.
+   * - {string} listingId - Unique listing ID.
+   * - {Object} chainOffer - Offer data from the blockchain.
+   * - {Object} ipfsOffer - Offer data from IPFS.
+   * @param {Object} discoveryData - Constructor argument used by discovery-service
+   * - {string} offerId - Unique offer ID.
+   * - {string} listingId - Unique listing ID.
+   * - {string} status - offer status
+   * - {string} buyerAddress - Buyer's address
+   * - {Object} discoveryData - Offer data from discovery server.
    */
-  constructor(offerId, listingId, chainOffer, ipfsOffer) {
-    this.id = offerId
-    this.listingId = listingId
-    this.status = chainOffer.status // 'created', 'accepted', 'disputed', 'finalized', 'sellerReviewed'
-    this.createdAt = chainOffer.createdAt // Time in seconds since epoch.
-    this.buyer = chainOffer.buyer
-    this.events = chainOffer.events
-    this.refund = chainOffer.refund
+  constructor(
+    originJsData,
+    discoveryData
+  ) {
+    if (originJsData && discoveryData)
+      throw 'Only one of constructor parameters should be present'
 
-    // See src/schemas/offer.json for fields stored in IPFS offer data.
-    Object.assign(this, ipfsOffer)
+    if (originJsData) {
+      const { offerId, listingId, chainOffer, ipfsOffer } = originJsData
+
+      this.id = offerId
+      this.listingId = listingId
+      this.status = chainOffer.status // 'created', 'accepted', 'disputed', 'finalized', 'sellerReviewed'
+      this.createdAt = chainOffer.createdAt // Time in seconds since epoch.
+      this.buyer = chainOffer.buyer
+      this.events = chainOffer.events
+      this.refund = chainOffer.refund
+
+      // See src/schemas/offer.json for fields stored in IPFS offer data.
+      Object.assign(this, ipfsOffer)
+    } else {
+      const { offerId, listingId, status, buyerAddress, discoveryData } = originJsData
```

Comment:
Actually looking at this code again I don't like it how model class is adjusting to the classes that are using it. It should be the other way around. Will change it so that the model class stays uniform and let the scripts that call that class prepare the data in the needed format

```
class Offer{
  constructor(id, listingId, status, createdAt, buyer, events, refund) {
      this.id = offerId
      this.listingId = listingId
      this.status = chainOffer.status // 'created', 'accepted', 'disputed', 'finalized', 'sellerReviewed'
      this.createdAt = chainOffer.createdAt // Time in seconds since epoch.
      this.buyer = chainOffer.buyer
      this.events = chainOffer.events
      this.refund = chainOffer.refund
  }
}
```